subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/DefaultTestFailure.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing;
+
+import org.gradle.api.tasks.testing.TestFailure;
+import org.gradle.api.tasks.testing.TestFailureDetails;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+
+public class DefaultTestFailure implements TestFailure {
+
+    private final Throwable rawFailure;
+    private final TestFailureDetails details;
+
+    public DefaultTestFailure(Throwable rawFailure, TestFailureDetails details) {
+        this.rawFailure = rawFailure;
+        this.details = details;
+    }
+
+    @Override
+    public Throwable getRawFailure() {
+        return rawFailure;
+    }
+
+    @Override
+    public TestFailureDetails getDetails() {
+        return details;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        DefaultTestFailure that = (DefaultTestFailure) o;
+
+        if (rawFailure != null ? !rawFailure.equals(that.rawFailure) : that.rawFailure != null) {
+            return false;
+        }
+        return details != null ? details.equals(that.details) : that.details == null;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = rawFailure != null ? rawFailure.hashCode() : 0;
+        result = 31 * result + (details != null ? details.hashCode() : 0);
+        return result;
+    }
+
+    public static TestFailure fromTestAssertionFailure(Throwable failure, String expected, String actual) {
+        DefaultTestFailureDetails details = new DefaultTestFailureDetails(failure.getMessage(), failure.getClass().getName(), stacktraceOf(failure), true, expected, actual);
+        return new DefaultTestFailure(failure, details);
+    }
+
+    public static TestFailure fromTestFrameworkFailure(Throwable failure) {
+        DefaultTestFailureDetails details = new DefaultTestFailureDetails(messageOf(failure), failure.getClass().getName(), stacktraceOf(failure), false, null, null);
+        return new DefaultTestFailure(failure, details);
+    }
+
```

Comment:
Could you please move these factory method to the `:enterprise` subproject so that the TD plugin can also use them?

Reply:
I don't see a dependency on the `:enterprise` subproject anywhere in other Gradle subprojects. Instead of moving, I assume I should simply copy factory methods to a new `TestFailures` class in that project. Can you confirm?

Reply:
@lptr Could you please chime in here since you recently did something similar for `PlaceholderExceptions`?

Reply:
What we need for the Enterprise plugin to access things is that they are exposed in `:enterprise` or one of the other `:enterprise-*` subprojects. I wouldn't copy these methods there, but add methods in `TestFailures` that would call `DefaultTestFailure.from*Failure()`, so the actual implementation of these methods is not duplicated.

Reply:
Fixed.

Reply:
Am I missing sth. or have new methods not been added to `:enterprise`, yet?

Reply:
My understanding was that I only need to expose the factory methods on the `TestFailure` interface and not move anything to the `enterprise` subproject.

Reply:
@lptr Was that what you had in mind?

Reply:
Ping